File: packages/core/src/integrated-router.ts
import { compareUnicodeCodePoints } from "./canonical.js";
import type { CompilerDiagnostic } from "./compiler.js";
import type { EdgeSpec, GraphSpec } from "./types.js";

export const ROUTER_CONDITION_API_VERSION =
  "graphengineering.reacher-z.github.io/pattern-conditions/v1alpha1" as const;

const ROUTE_ID = /^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/;
const LOOP_CONDITIONS = new Set([
  "LoopContinue",
  "LoopDryVerdict",
  "LoopVerdictAtBound",
]);
const POLICY_KEYS = new Set([
  "kind",
  "allowedRoutes",
  "defaultRoute",
  "confidence",
  "maxMulticast",
]);
const CONFIDENCE_KEYS = new Set(["minimumBasisPoints", "escalationRoute"]);
const ROUTE_CONDITION_KEYS = new Set(["apiVersion", "kind", "routeKey"]);

export interface RouteSelectionPolicySnapshot {
  readonly kind: "single" | "multi";
  readonly allowedRoutes: readonly string[];
  readonly defaultRoute?: string;
  readonly confidence?: Readonly<{
    minimumBasisPoints: number;
    escalationRoute: string;
  }>;
  readonly maxMulticast?: number;
}

export interface InvalidRouterValue {
  readonly valid: false;
  readonly relativePath: string;
}

export interface ValidRouterPolicy {
  readonly valid: true;
  readonly policy: RouteSelectionPolicySnapshot;
}

export type ValidRegisteredCondition =
  | { readonly valid: true; readonly owner: "loop-pattern" }
  | { readonly valid: true; readonly owner: "integrated-router"; readonly routeKey: string };

export type PolicyValidation = InvalidRouterValue | ValidRouterPolicy;

export type ConditionValidation = InvalidRouterValue | ValidRegisteredCondition;

function record(value: unknown): value is Readonly<Record<string, unknown>> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function routeId(value: unknown): value is string {
  return typeof value === "string"
    && ROUTE_ID.test(value)
    && value !== "."
    && value !== "..";
}

function firstUnknown(
  value: Readonly<Record<string, unknown>>,
  allowed: ReadonlySet<string>,
): string | undefined {
  return Object.keys(value)
    .filter((key) => !allowed.has(key))
    .sort(compareUnicodeCodePoints)[0];
}

export function validateRouteSelectionPolicy(value: unknown): PolicyValidation {
  if (!record(value)) return { valid: false, relativePath: "" };
  const unknown = firstUnknown(value, POLICY_KEYS);
  if (unknown !== undefined) return { valid: false, relativePath: `/${unknown}` };

  if (value.kind !== "single" && value.kind !== "multi") {
    return { valid: false, relativePath: "/kind" };
  }
  if (!Array.isArray(value.allowedRoutes) || value.allowedRoutes.length === 0) {
    return { valid: false, relativePath: "/allowedRoutes" };
  }
  const allowedRoutes: string[] = [];
  const seen = new Set<string>();
  for (const [index, candidate] of value.allowedRoutes.entries()) {
    if (!routeId(candidate) || seen.has(candidate)) {
      return { valid: false, relativePath: `/allowedRoutes/${index}` };
    }
    seen.add(candidate);
    allowedRoutes.push(candidate);
  }

  if (Object.hasOwn(value, "defaultRoute")
      && (!routeId(value.defaultRoute) || !seen.has(value.defaultRoute))) {
    return { valid: false, relativePath: "/defaultRoute" };
  }

  let confidence: RouteSelectionPolicySnapshot["confidence"];
  if (Object.hasOwn(value, "confidence")) {
    if (!record(value.confidence)) {
      return { valid: false, relativePath: "/confidence" };
    }
    const confidenceUnknown = firstUnknown(value.confidence, CONFIDENCE_KEYS);
    if (confidenceUnknown !== undefined) {
      return { valid: false, relativePath: `/confidence/${confidenceUnknown}` };
    }
    if (!Number.isSafeInteger(value.confidence.minimumBasisPoints)
        || (value.confidence.minimumBasisPoints as number) < 1
        || (value.confidence.minimumBasisPoints as number) > 10_000) {
      return { valid: false, relativePath: "/confidence/minimumBasisPoints" };
    }
    if (!routeId(value.confidence.escalationRoute)
        || !seen.has(value.confidence.escalationRoute)) {
      return { valid: false, relativePath: "/confidence/escalationRoute" };
    }
    confidence = Object.freeze({
      minimumBasisPoints: value.confidence.minimumBasisPoints as number,
      escalationRoute: value.confidence.escalationRoute,
    });
  }

  let maxMulticast: number | undefined;
  if (value.kind === "single") {
    if (Object.hasOwn(value, "maxMulticast")) {
      return { valid: false, relativePath: "/maxMulticast" };
    }
  } else {
    if (!Number.isSafeInteger(value.maxMulticast)
        || (value.maxMulticast as number) < 1
        || (value.maxMulticast as number) > allowedRoutes.length) {
      return { valid: false, relativePath: "/maxMulticast" };
    }
    maxMulticast = value.maxMulticast as number;
  }

  return {
    valid: true,
    policy: Object.freeze({
      kind: value.kind,
      allowedRoutes: Object.freeze(allowedRoutes),
      ...(Object.hasOwn(value, "defaultRoute")
        ? { defaultRoute: value.defaultRoute as string }
        : {}),
      ...(confidence === undefined ? {} : { confidence }),
      ...(maxMulticast === undefined ? {} : { maxMulticast }),
    }),
  };
}

export function validateRegisteredEdgeCondition(value: unknown): ConditionValidation {
  if (!record(value)) return { valid: false, relativePath: "" };
  if (!Object.hasOwn(value, "apiVersion") || typeof value.apiVersion !== "string") {
    return { valid: false, relativePath: "/apiVersion" };
  }
  if (!Object.hasOwn(value, "kind") || typeof value.kind !== "string") {
    return { valid: false, relativePath: "/kind" };
  }
  if (value.apiVersion !== ROUTER_CONDITION_API_VERSION) {
    return { valid: false, relativePath: "/apiVersion" };
  }
  if (LOOP_CONDITIONS.has(value.kind)) {
    return { valid: true, owner: "loop-pattern" };
  }
  if (value.kind !== "RouteEquals") {
    return { valid: false, relativePath: "/kind" };
  }
  const unknown = firstUnknown(value, ROUTE_CONDITION_KEYS);
  if (unknown !== undefined) return { valid: false, relativePath: `/${unknown}` };
  if (!routeId(value.routeKey)) {
    return { valid: false, relativePath: "/routeKey" };
  }
  return { valid: true, owner: "integrated-router", routeKey: value.routeKey };
}

function diagnostic(
  code:
    | "GE1401_INVALID_ROUTER_POLICY"
    | "GE1402_UNSUPPORTED_EDGE_CONDITION"
    | "GE1403_CONDITION_SOURCE_NOT_ROUTER"
    | "GE1404_ROUTE_NOT_ALLOWED"
    | "GE1405_DUPLICATE_ROUTE_CASE"
    | "GE1406_DUPLICATE_ROUTE_TARGET"
    | "GE1407_INCOMPLETE_ROUTE_COVERAGE",
  message: string,
  path: string,
  fields: Pick<CompilerDiagnostic, "nodeIds" | "edgeId">,
): CompilerDiagnostic {
  return Object.freeze({
    code,
    severity: "error",
    message,
    path,
    nodeIds: Object.freeze([...(fields.nodeIds ?? [])]),
    ...(fields.edgeId === undefined ? {} : { edgeId: fields.edgeId }),
  });
}

interface RouteCandidate {
  readonly edge: EdgeSpec;
  readonly edgeIndex: number;
  readonly routerId: string;
  readonly routeKey: string;
}

/** Validate integrated router policy and RouteEquals topology on a captured graph. */
export function validateIntegratedRouterSnapshot(
  graph: GraphSpec,
): readonly CompilerDiagnostic[] {
  const policies = new Map<string, RouteSelectionPolicySnapshot>();
  const invalidRouters = new Set<string>();
  const nodesById = new Map(graph.nodes.map((node) => [node.id, node]));
  const nodeIndex = new Map(graph.nodes.map((node, index) => [node.id, index]));
  const policyDiagnostics: CompilerDiagnostic[] = [];
  for (const [index, node] of graph.nodes.entries()) {
    if (node.kind !== "router") continue;
    const validated = validateRouteSelectionPolicy(node.config);
    if (!validated.valid) {
      invalidRouters.add(node.id);
      policyDiagnostics.push(diagnostic(
        "GE1401_INVALID_ROUTER_POLICY",
        `Router '${node.id}' has an invalid route-selection policy`,
        `#/nodes/${index}/config${validated.relativePath}`,
        { nodeIds: [node.id] },
      ));
    } else {
      policies.set(node.id, validated.policy);
    }
  }

  const unsupportedDiagnostics: CompilerDiagnostic[] = [];
  const sourceDiagnostics: CompilerDiagnostic[] = [];
  const candidates: RouteCandidate[] = [];
  const suppressCoverage = new Set<string>();
  for (const [index, edge] of graph.edges.entries()) {
    if (edge.condition === undefined) continue;
    const validated = validateRegisteredEdgeCondition(edge.condition);
    if (!validated.valid) {
      unsupportedDiagnostics.push(diagnostic(
        "GE1402_UNSUPPORTED_EDGE_CONDITION",
        `Edge '${edge.id}' has an unsupported or malformed condition`,
        `#/edges/${index}/condition${validated.relativePath}`,
        { edgeId: edge.id, nodeIds: [edge.from.node] },
      ));
      if (policies.has(edge.from.node)) suppressCoverage.add(edge.from.node);
      continue;
    }
    if (validated.owner === "loop-pattern") continue;
    const source = nodesById.get(edge.from.node);
    if (source?.kind !== "router") {
      sourceDiagnostics.push(diagnostic(
        "GE1403_CONDITION_SOURCE_NOT_ROUTER",
        `Edge '${edge.id}' uses RouteEquals but source '${edge.from.node}' is not a router`,
        `#/edges/${index}/condition`,
        { edgeId: edge.id, nodeIds: [edge.from.node] },
      ));
      continue;
    }
    if (invalidRouters.has(source.id)) continue;
    candidates.push({
      edge,
      edgeIndex: index,
      routerId: source.id,
      routeKey: validated.routeKey,
    });
  }

  const notAllowedDiagnostics: CompilerDiagnostic[] = [];
  const duplicateCaseDiagnostics: CompilerDiagnostic[] = [];
  const duplicateTargetDiagnostics: CompilerDiagnostic[] = [];
  const allowedRoutes = new Map(
    [...policies].map(([routerId, policy]) => [routerId, new Set(policy.allowedRoutes)]),
  );
  const seenRoutes = new Map([...policies].map(([routerId]) => [routerId, new Set<string>()]));
  const seenTargets = new Map([...policies].map(([routerId]) => [routerId, new Set<string>()]));
  const acceptedRoutes = new Map([...policies].map(([routerId]) => [routerId, new Set<string>()]));
  for (const candidate of candidates) {
    const { edge, edgeIndex, routerId, routeKey } = candidate;
    const allowed = allowedRoutes.get(routerId)!;
    const routerSeenRoutes = seenRoutes.get(routerId)!;
    const routerSeenTargets = seenTargets.get(routerId)!;
    const accepted = acceptedRoutes.get(routerId)!;
    if (!allowed.has(routeKey)) {
      suppressCoverage.add(routerId);
      notAllowedDiagnostics.push(diagnostic(
        "GE1404_ROUTE_NOT_ALLOWED",
        `Edge '${edge.id}' route '${routeKey}' is not allowed by router '${routerId}'`,
        `#/edges/${edgeIndex}/condition/routeKey`,
        { edgeId: edge.id, nodeIds: [routerId] },
      ));
      continue;
    }
    if (routerSeenRoutes.has(routeKey)) {
      suppressCoverage.add(routerId);
      duplicateCaseDiagnostics.push(diagnostic(
        "GE1405_DUPLICATE_ROUTE_CASE",
        `Edge '${edge.id}' repeats route '${routeKey}' for router '${routerId}'`,
        `#/edges/${edgeIndex}/condition/routeKey`,
        { edgeId: edge.id, nodeIds: [routerId] },
      ));
      continue;
    }
    routerSeenRoutes.add(routeKey);
    accepted.add(routeKey);
    if (routerSeenTargets.has(edge.to.node)) {
      suppressCoverage.add(routerId);
      duplicateTargetDiagnostics.push(diagnostic(
        "GE1406_DUPLICATE_ROUTE_TARGET",
        `Edge '${edge.id}' repeats target '${edge.to.node}' for router '${routerId}'`,
        `#/edges/${edgeIndex}/to/node`,
        { edgeId: edge.id, nodeIds: [routerId, edge.to.node] },
      ));
    } else {
      routerSeenTargets.add(edge.to.node);
    }
  }

  const coverageDiagnostics: CompilerDiagnostic[] = [];
  for (const node of graph.nodes) {
    const policy = policies.get(node.id);
    if (policy === undefined || suppressCoverage.has(node.id)) continue;
    const accepted = acceptedRoutes.get(node.id) ?? new Set<string>();
    const missing = policy.allowedRoutes.filter((route) => !accepted.has(route));
    if (missing.length === 0) continue;
    coverageDiagnostics.push(diagnostic(
      "GE1407_INCOMPLETE_ROUTE_COVERAGE",
      `Router '${node.id}' is missing route cases: ${missing.join(", ")}`,
      `#/nodes/${nodeIndex.get(node.id) ?? 0}/config/allowedRoutes`,
      { nodeIds: [node.id] },
    ));
  }

  return Object.freeze([
    ...policyDiagnostics,
    ...unsupportedDiagnostics,
    ...sourceDiagnostics,
    ...notAllowedDiagnostics,
    ...duplicateCaseDiagnostics,
    ...duplicateTargetDiagnostics,
    ...coverageDiagnostics,
  ]);
}

File: python/src/graph_engineering/integrated_router.py
"""Integrated router policy and RouteEquals compiler validation."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal, TypeAlias, TypeGuard

from .compiler import Diagnostic, DiagnosticCode
from .models import MAX_SAFE_INTEGER, EdgeSpec, GraphSpec, JsonValue

ROUTER_CONDITION_API_VERSION = "graphengineering.reacher-z.github.io/pattern-conditions/v1alpha1"

_ROUTE_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_LOOP_CONDITIONS = frozenset({"LoopContinue", "LoopDryVerdict", "LoopVerdictAtBound"})
_POLICY_KEYS = frozenset({"kind", "allowedRoutes", "defaultRoute", "confidence", "maxMulticast"})
_CONFIDENCE_KEYS = frozenset({"minimumBasisPoints", "escalationRoute"})
_ROUTE_CONDITION_KEYS = frozenset({"apiVersion", "kind", "routeKey"})


@dataclass(frozen=True, slots=True)
class RouteSelectionConfidenceSnapshot:
    minimum_basis_points: int
    escalation_route: str


@dataclass(frozen=True, slots=True)
class RouteSelectionPolicySnapshot:
    kind: Literal["single", "multi"]
    allowed_routes: tuple[str, ...]
    default_route: str | None = None
    confidence: RouteSelectionConfidenceSnapshot | None = None
    max_multicast: int | None = None


@dataclass(frozen=True, slots=True)
class InvalidRouterValue:
    valid: Literal[False]
    relative_path: str


@dataclass(frozen=True, slots=True)
class ValidRouterPolicy:
    valid: Literal[True]
    policy: RouteSelectionPolicySnapshot


@dataclass(frozen=True, slots=True)
class ValidRegisteredCondition:
    valid: Literal[True]
    owner: Literal["integrated-router", "loop-pattern"]
    route_key: str | None = None


PolicyValidation: TypeAlias = InvalidRouterValue | ValidRouterPolicy
ConditionValidation: TypeAlias = InvalidRouterValue | ValidRegisteredCondition


def _record(value: object) -> TypeGuard[dict[str, JsonValue]]:
    return type(value) is dict


def _route_id(value: object) -> TypeGuard[str]:
    if type(value) is not str:
        return False
    return _ROUTE_ID.fullmatch(value) is not None and value not in {".", ".."}


def _safe_integer(value: object) -> TypeGuard[int | float]:
    if type(value) is int:
        return abs(value) <= MAX_SAFE_INTEGER
    return type(value) is float and value.is_integer() and abs(value) <= MAX_SAFE_INTEGER


def _first_unknown(value: dict[str, JsonValue], allowed: frozenset[str]) -> str | None:
    return next(iter(sorted(key for key in value if key not in allowed)), None)


def validate_route_selection_policy(value: object) -> PolicyValidation:
    """Validate one direct exact router config without inferring from edges."""

    if not _record(value):
        return InvalidRouterValue(False, "")
    record = value
    unknown = _first_unknown(record, _POLICY_KEYS)
    if unknown is not None:
        return InvalidRouterValue(False, f"/{unknown}")

    kind = record.get("kind")
    if kind not in {"single", "multi"}:
        return InvalidRouterValue(False, "/kind")
    policy_kind: Literal["single", "multi"] = "single" if kind == "single" else "multi"
    routes_value = record.get("allowedRoutes")
    if type(routes_value) is not list or not routes_value:
        return InvalidRouterValue(False, "/allowedRoutes")
    allowed_routes: list[str] = []
    seen: set[str] = set()
    for index, candidate in enumerate(routes_value):
        if not _route_id(candidate) or candidate in seen:
            return InvalidRouterValue(False, f"/allowedRoutes/{index}")
        seen.add(candidate)
        allowed_routes.append(candidate)

    default_route: str | None = None
    if "defaultRoute" in record:
        candidate = record["defaultRoute"]
        if not _route_id(candidate) or candidate not in seen:
            return InvalidRouterValue(False, "/defaultRoute")
        default_route = candidate

    confidence: RouteSelectionConfidenceSnapshot | None = None
    if "confidence" in record:
        confidence_value = record["confidence"]
        if not _record(confidence_value):
            return InvalidRouterValue(False, "/confidence")
        confidence_record = confidence_value
        confidence_unknown = _first_unknown(confidence_record, _CONFIDENCE_KEYS)
        if confidence_unknown is not None:
            return InvalidRouterValue(False, f"/confidence/{confidence_unknown}")
        minimum = confidence_record.get("minimumBasisPoints")
        if not _safe_integer(minimum) or not 1 <= minimum <= 10_000:
            return InvalidRouterValue(False, "/confidence/minimumBasisPoints")
        escalation = confidence_record.get("escalationRoute")
        if not _route_id(escalation) or escalation not in seen:
            return InvalidRouterValue(False, "/confidence/escalationRoute")
        confidence = RouteSelectionConfidenceSnapshot(int(minimum), escalation)

    max_multicast: int | None = None
    if policy_kind == "single":
        if "maxMulticast" in record:
            return InvalidRouterValue(False, "/maxMulticast")
    else:
        maximum = record.get("maxMulticast")
        if not _safe_integer(maximum) or not 1 <= maximum <= len(allowed_routes):
            return InvalidRouterValue(False, "/maxMulticast")
        max_multicast = int(maximum)

    return ValidRouterPolicy(
        True,
        RouteSelectionPolicySnapshot(
            policy_kind,
            tuple(allowed_routes),
            default_route,
            confidence,
            max_multicast,
        ),
    )


def validate_registered_edge_condition(value: object) -> ConditionValidation:
    """Classify the exact shared condition registry and validate RouteEquals."""

    if not _record(value):
        return InvalidRouterValue(False, "")
    record = value
    if "apiVersion" not in record or type(record["apiVersion"]) is not str:
        return InvalidRouterValue(False, "/apiVersion")
    if "kind" not in record or type(record["kind"]) is not str:
        return InvalidRouterValue(False, "/kind")
    if record["apiVersion"] != ROUTER_CONDITION_API_VERSION:
        return InvalidRouterValue(False, "/apiVersion")
    if record["kind"] in _LOOP_CONDITIONS:
        return ValidRegisteredCondition(True, "loop-pattern")
    if record["kind"] != "RouteEquals":
        return InvalidRouterValue(False, "/kind")
    unknown = _first_unknown(record, _ROUTE_CONDITION_KEYS)
    if unknown is not None:
        return InvalidRouterValue(False, f"/{unknown}")
    route_key = record.get("routeKey")
    if not _route_id(route_key):
        return InvalidRouterValue(False, "/routeKey")
    return ValidRegisteredCondition(True, "integrated-router", route_key)


@dataclass(frozen=True, slots=True)
class _RouteCandidate:
    edge: EdgeSpec
    edge_index: int
    router_id: str
    route_key: str


def _diagnostic(
    code: DiagnosticCode,
    message: str,
    path: str,
    node_ids: tuple[str, ...],
    edge_id: str | None = None,
) -> Diagnostic:
    return Diagnostic(
        code=code,
        message=message,
        node_id=node_ids[0] if node_ids else None,
        edge_id=edge_id,
        path=path,
        node_ids=node_ids,
    )


def validate_integrated_router_snapshot(graph: GraphSpec) -> tuple[Diagnostic, ...]:
    """Validate router policies and route-table topology in frozen pass order."""

    policies: dict[str, RouteSelectionPolicySnapshot] = {}
    invalid_routers: set[str] = set()
    nodes_by_id = {node.id: node for node in graph.nodes}
    node_index = {node.id: index for index, node in enumerate(graph.nodes)}
    policy_diagnostics: list[Diagnostic] = []
    for index, node in enumerate(graph.nodes):
        if node.kind != "router":
            continue
        policy_validation = validate_route_selection_policy(node.config)
        if not policy_validation.valid:
            invalid_routers.add(node.id)
            policy_diagnostics.append(
                _diagnostic(
                    DiagnosticCode.INVALID_ROUTER_POLICY,
                    f"router {node.id!r} has an invalid route-selection policy",
                    f"#/nodes/{index}/config{policy_validation.relative_path}",
                    (node.id,),
                )
            )
        else:
            policies[node.id] = policy_validation.policy

    unsupported_diagnostics: list[Diagnostic] = []
    source_diagnostics: list[Diagnostic] = []
    candidates: list[_RouteCandidate] = []
    suppress_coverage: set[str] = set()
    for index, edge in enumerate(graph.edges):
        if edge.condition is None:
            continue
        condition_validation = validate_registered_edge_condition(edge.condition)
        if not condition_validation.valid:
            unsupported_diagnostics.append(
                _diagnostic(
                    DiagnosticCode.UNSUPPORTED_EDGE_CONDITION,
                    f"edge {edge.id!r} has an unsupported or malformed condition",
                    f"#/edges/{index}/condition{condition_validation.relative_path}",
                    (edge.source.node,),
                    edge.id,
                )
            )
            if edge.source.node in policies:
                suppress_coverage.add(edge.source.node)
            continue
        if condition_validation.owner == "loop-pattern":
            continue
        source = nodes_by_id.get(edge.source.node)
        if source is None or source.kind != "router":
            source_diagnostics.append(
                _diagnostic(
                    DiagnosticCode.CONDITION_SOURCE_NOT_ROUTER,
                    f"edge {edge.id!r} uses RouteEquals from non-router {edge.source.node!r}",
                    f"#/edges/{index}/condition",
                    (edge.source.node,),
                    edge.id,
                )
            )
            continue
        if source.id in invalid_routers:
            continue
        if condition_validation.route_key is None:
            raise AssertionError("integrated-router condition omitted its route key")
        candidates.append(_RouteCandidate(edge, index, source.id, condition_validation.route_key))

    not_allowed_diagnostics: list[Diagnostic] = []
    duplicate_case_diagnostics: list[Diagnostic] = []
    duplicate_target_diagnostics: list[Diagnostic] = []
    seen_routes: dict[str, set[str]] = {router_id: set() for router_id in policies}
    seen_targets: dict[str, set[str]] = {router_id: set() for router_id in policies}
    accepted_routes: dict[str, set[str]] = {router_id: set() for router_id in policies}
    for candidate in candidates:
        candidate_policy = policies[candidate.router_id]
        edge = candidate.edge
        if candidate.route_key not in candidate_policy.allowed_routes:
            suppress_coverage.add(candidate.router_id)
            not_allowed_diagnostics.append(
                _diagnostic(
                    DiagnosticCode.ROUTE_NOT_ALLOWED,
                    f"edge {edge.id!r} route {candidate.route_key!r} is not allowed",
                    f"#/edges/{candidate.edge_index}/condition/routeKey",
                    (candidate.router_id,),
                    edge.id,
                )
            )
            continue
        if candidate.route_key in seen_routes[candidate.router_id]:
            suppress_coverage.add(candidate.router_id)
            duplicate_case_diagnostics.append(
                _diagnostic(
                    DiagnosticCode.DUPLICATE_ROUTE_CASE,
                    f"edge {edge.id!r} repeats route {candidate.route_key!r}",
                    f"#/edges/{candidate.edge_index}/condition/routeKey",
                    (candidate.router_id,),
                    edge.id,
                )
            )
            continue
        seen_routes[candidate.router_id].add(candidate.route_key)
        accepted_routes[candidate.router_id].add(candidate.route_key)
        target = edge.target.node
        if target in seen_targets[candidate.router_id]:
            suppress_coverage.add(candidate.router_id)
            duplicate_target_diagnostics.append(
                _diagnostic(
                    DiagnosticCode.DUPLICATE_ROUTE_TARGET,
                    f"edge {edge.id!r} repeats target {target!r}",
                    f"#/edges/{candidate.edge_index}/to/node",
                    (candidate.router_id, target),
                    edge.id,
                )
            )
        else:
            seen_targets[candidate.router_id].add(target)

    coverage_diagnostics: list[Diagnostic] = []
    for node in graph.nodes:
        coverage_policy = policies.get(node.id)
        if coverage_policy is None or node.id in suppress_coverage:
            continue
        accepted = accepted_routes[node.id]
        missing = tuple(route for route in coverage_policy.allowed_routes if route not in accepted)
        if not missing:
            continue
        coverage_diagnostics.append(
            _diagnostic(
                DiagnosticCode.INCOMPLETE_ROUTE_COVERAGE,
                f"router {node.id!r} is missing route cases: {', '.join(missing)}",
                f"#/nodes/{node_index[node.id]}/config/allowedRoutes",
                (node.id,),
            )
        )

    return (
        *policy_diagnostics,
        *unsupported_diagnostics,
        *source_diagnostics,
        *not_allowed_diagnostics,
        *duplicate_case_diagnostics,
        *duplicate_target_diagnostics,
        *coverage_diagnostics,
    )
